### PR TITLE
Add reproy process for samplify request.

### PR DESCRIPTION
**Pull request by saratily**

### CODE REVIEW

**Review by caseylucas — COMMENTED** [review]
I noticed that some of the public func are getting parameter changes.  This of course would be a breaking change for this public lib.  CC @manikanta-talanki @klirot

**Comment by saratily** [comment]
> I noticed that some of the public func are getting parameter changes. This of course would be a breaking change for this public lib. CC @manikanta-talanki @klirot

I refactor the code, now there is no change in public func. @caseylucas Could you please review again

File: lib/client.go [hunk 1]
```diff
@@ -939,6 +942,11 @@ func NewClient(clientID, username, passsword string, options *ClientOptions) *Cl
 		options.Timeout = &timeout
 	}
 
+	if options != nil && options.MaxRetries == 0 {
+		retryMax := defaultretrymax
+		options.MaxRetries = retryMax
```
Comment (manikanta-talanki) [inline lib/client.go:947]: why not ?
```go
options.MaxRetries = defaultretrymax
```

File: lib/client.go [hunk 2]
```diff
@@ -950,7 +958,7 @@ func NewClient(clientID, username, passsword string, options *ClientOptions) *Cl
 }
 
 // SetOptions ...
-func (c *Client) SetOptions(env string, timeout int) error {
+func (c *Client) SetOptions(env string, timeout, maxRetries int) error {
```
Comment (manikanta-talanki) [inline lib/client.go:961]: Do not change the signature of an Exported function. Use the Options pattern.

File: lib/client.go [hunk 3]
```diff
@@ -972,19 +980,25 @@ func (c *Client) SetOptions(env string, timeout int) error {
...
+	}
+
+	c.Options.MaxRetries = maxRetries
+
 	return nil
 }
 
 // NewClientFromEnv returns an API client.
-func NewClientFromEnv(clientID, username, passsword string, env string, timeout int) (*Client, error) {
+func NewClientFromEnv(clientID, username, passsword string, env string, timeout, retryMax int) (*Client, error) {
```
Comment (manikanta-talanki) [inline lib/client.go:993]: Same here, Do not change the signatures of Exported functions.

File: lib/request.go [hunk 4]
```diff
@@ -19,25 +21,22 @@ type APIResponse struct {
 }
 
 // SendRequestWithContext exposing sendrequest to enable custom requests
-func SendRequestWithContext(ctx context.Context, host, method, url, accessToken string, body interface{}, timeout int) (*APIResponse, error) {
-	return sendRequest(ctx, host, method, url, accessToken, body, timeout)
+func SendRequestWithContext(ctx context.Context, host, method, url, accessToken string, body interface{}, co *ClientOptions) (*APIResponse, error) {
```
Comment (manikanta-talanki) [inline lib/request.go:24]: here too.

File: lib/request.go [hunk 5]
```diff
@@ -48,7 +47,28 @@ func sendRequest(ctx context.Context, host, method, url, accessToken string, bod
 	if len(accessToken) > 0 {
 		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", accessToken))
 	}
-	resp, err := client.Do(req)
+
+	if err != nil {
+		panic(err)
+	}
```
Comment (manikanta-talanki) [inline lib/request.go:53]: Is this necessary? and please do not use `panic` particularly in a library.

File: lib/request.go [hunk 6]
```diff
@@ -48,7 +47,28 @@ func sendRequest(ctx context.Context, host, method, url, accessToken string, bod
...
+	retryableClient := retryablehttp.NewClient()
+
+	dur := time.Duration(*co.Timeout)
+	retryableClient.HTTPClient = &http.Client{
+		Timeout: time.Second * dur,
+	}
```
Comment (manikanta-talanki) [inline lib/request.go:59]: whats the intent here?

**Review by manikanta-talanki — CHANGES_REQUESTED** [review]

**Comment by saratily** [comment]
Use github.com/hashicorp/go-retryablehttp for retry process

**Pull request closed without merging**